multidict/_multidict.c

```diff
@@ -135,28 +135,40 @@ _multidict_append_items_seq(MultiDictObject *self, PyObject *arg,
     }
 
     while ((item = PyIter_Next(iter)) != NULL) {
-        if (PyObject_Length(item) != 2) {
-            PyErr_Format(
-                PyExc_TypeError,
-                "%s takes either dict or list of (key, value) tuples",
-                name,
-                NULL
-            );
-            Py_DECREF(item);
-            Py_DECREF(iter);
-            return -1;
+        if (PyTuple_CheckExact(item)) {
+            if (PyTuple_GET_SIZE(item) != 2) {
+                goto invalid_type;
+            }
+            key   = PyTuple_GET_ITEM(item, 0);
+            Py_INCREF(key);
```

Comment:
This is a potential memleak because we increment key and val ref in pair_list_add -> _pair_list_add_with_hash.

Reply:
I missed `Py_CLEAR` at the end of the loop. Ok, but I don't think that we need `Py_INCREF` and `Py_CLEAR` for key and val, `pair_list_add` do this job.

Reply:
`key` and `value` are decrefed at the end of `while` loop, see `Py_CLEAR()` calls.

Bugs on CPython taught me that borrowed references are things that should be avoided.
That's hard but still possible to write a custom `key` class in pure Python that will crash the interpreter when used with a multidict.  Say, it can destroy a list in its own `__hash__` function.
Holding a reference until `pair_list_add` is called prevents this weird-but-still-possible bug.